### PR TITLE
Create resources with cloudformation (fixes #90)

### LINKED ISSUE #90: Refactor to deploy resources in the serverless cloudformation stack
The current setup with having additional commands to execute outside of a typical `sls deploy` that create resources that aren't tied to the sls CloudFormation stack is problematic for a number of reasons:
* Non-abstracted dependency and versioning/rollback issues of having to execute `sls deploy` and `sls appsync-deploy` or `sls appsync-update` sequentially
* Requirement of updating configuration after performing a deploy to inject the `apiId`
* Having to know whether to call update or deploy
* Isolated resources cannot be easily cleaned up with the rest of serverless stack
* Obfuscation of what resources this plugin is actually creating
* Inability to further customize the generated resources as new features are added to AppSync

I believe this change would also address these issues:
* #76 
* #81
* #49 

I plan on implementing this myself for a project I'm working on now and I'd like to contribute back. @sid88in Would you be open to accepting a PR that implements this?

We could implement this in one of two ways, breaking or non-breaking.  I think the breaking change would be recommended as we want to enforce this behavior by default going forward and reduce accumulation of technical debt.

For the breaking change, I would propose this acceptance criteria:
* Compatibility with existing configuration block with the exception of apiId
* Hooks into `before:deploy:deploy` lifecycle event to add AppSync resources to  `serverless.service.provider.compiledCloudFormationTemplate.Resources`
* Log and return an error when `appsync-update` or `appsync-deploy` is called

### PR DESCRIPTION
**Pull request by ajmath**
Fixes #90, #76, #81, #49 

This was a straight replacement of the AWS api calls to create/update AppSync resources with adding the Cloudformation equivalent to the serverless stack's resources.

The previous cli commands have been updated to throw an exception with a pointer to a migration section of the README.  The exception is the `delete-appsync` command which remains to aid in the migration effort.

Here is the output from `sls deploy` in the example:
```shell
$ sls deploy --verbose
Serverless: Removing /Users/amatheny/code/serverless-appsync-plugin/example/.webpack
Serverless: Bundling with Webpack...
Time: 395ms
     Asset     Size  Chunks             Chunk Names
handler.js  11.2 kB       0  [emitted]  handler
   [0] ./handler.js 8.37 kB {0} [built]
   [1] external "babel-polyfill" 42 bytes {0} [not cacheable]
   [2] external "oauth" 42 bytes {0} [not cacheable]
   [3] external "twitter" 42 bytes {0} [not cacheable]
Serverless: Zip service: /Users/amatheny/code/serverless-appsync-plugin/example/.webpack/service [15 ms]
Serverless: Packaging service...
Serverless: Remove /Users/amatheny/code/serverless-appsync-plugin/example/.webpack
Serverless: Creating Stack...
Serverless: Checking Stack create progress...
CloudFormation - CREATE_IN_PROGRESS - AWS::CloudFormation::Stack - appsync-example-dev
CloudFormation - CREATE_IN_PROGRESS - AWS::S3::Bucket - ServerlessDeploymentBucket
CloudFormation - CREATE_IN_PROGRESS - AWS::S3::Bucket - ServerlessDeploymentBucket
CloudFormation - CREATE_COMPLETE - AWS::S3::Bucket - ServerlessDeploymentBucket
CloudFormation - CREATE_COMPLETE - AWS::CloudFormation::Stack - appsync-example-dev
Serverless: Stack create finished...
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Uploading service .zip file to S3 (2.81 KB)...
Serverless: Validating template...
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
CloudFormation - UPDATE_IN_PROGRESS - AWS::CloudFormation::Stack - appsync-example-dev
CloudFormation - CREATE_IN_PROGRESS - AWS::DynamoDB::Table - UserTable
CloudFormation - CREATE_IN_PROGRESS - AWS::IAM::Role - AppSyncESServiceRole
CloudFormation - CREATE_IN_PROGRESS - AWS::IAM::Role - AppSyncDynamoDBServiceRole
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::GraphQLApi - GraphQlApi
CloudFormation - CREATE_IN_PROGRESS - AWS::IAM::Role - IamRoleLambdaExecution
CloudFormation - CREATE_IN_PROGRESS - AWS::Logs::LogGroup - GraphqlLogGroup
CloudFormation - CREATE_IN_PROGRESS - AWS::DynamoDB::Table - TweetsTable
CloudFormation - CREATE_IN_PROGRESS - AWS::Logs::LogGroup - GraphqlLogGroup
CloudFormation - CREATE_IN_PROGRESS - AWS::DynamoDB::Table - UserTable
CloudFormation - CREATE_IN_PROGRESS - AWS::IAM::Role - AppSyncLambdaServiceRole
CloudFormation - CREATE_IN_PROGRESS - AWS::DynamoDB::Table - TweetsTable
CloudFormation - CREATE_COMPLETE - AWS::Logs::LogGroup - GraphqlLogGroup
CloudFormation - CREATE_IN_PROGRESS - AWS::IAM::Role - AppSyncDynamoDBServiceRole
CloudFormation - CREATE_IN_PROGRESS - AWS::IAM::Role - AppSyncESServiceRole
CloudFormation - CREATE_IN_PROGRESS - AWS::IAM::Role - IamRoleLambdaExecution
CloudFormation - CREATE_IN_PROGRESS - AWS::IAM::Role - AppSyncLambdaServiceRole
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::GraphQLApi - GraphQlApi
CloudFormation - CREATE_COMPLETE - AWS::AppSync::GraphQLApi - GraphQlApi
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::GraphQLSchema - GraphQlSchema
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::GraphQLSchema - GraphQlSchema
CloudFormation - CREATE_COMPLETE - AWS::IAM::Role - AppSyncESServiceRole
CloudFormation - CREATE_COMPLETE - AWS::IAM::Role - AppSyncDynamoDBServiceRole
CloudFormation - CREATE_COMPLETE - AWS::IAM::Role - AppSyncLambdaServiceRole
CloudFormation - CREATE_COMPLETE - AWS::IAM::Role - IamRoleLambdaExecution
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::DataSource - GraphQlDsElasticSearch
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Function - GraphqlLambdaFunction
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::DataSource - GraphQlDsElasticSearch
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Function - GraphqlLambdaFunction
CloudFormation - CREATE_COMPLETE - AWS::AppSync::DataSource - GraphQlDsElasticSearch
CloudFormation - CREATE_COMPLETE - AWS::Lambda::Function - GraphqlLambdaFunction
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Version - GraphqlLambdaVersionc0EfMITK4HKsUojkEJZj5f33THE6rv2EoDwoXWppZk
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Version - GraphqlLambdaVersionc0EfMITK4HKsUojkEJZj5f33THE6rv2EoDwoXWppZk
CloudFormation - CREATE_COMPLETE - AWS::Lambda::Version - GraphqlLambdaVersionc0EfMITK4HKsUojkEJZj5f33THE6rv2EoDwoXWppZk
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::DataSource - GraphQlDsLambdaMeInfo
CloudFormation - CREATE_COMPLETE - AWS::AppSync::GraphQLSchema - GraphQlSchema
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::DataSource - GraphQlDsLambdaMeInfo
CloudFormation - CREATE_COMPLETE - AWS::AppSync::DataSource - GraphQlDsLambdaMeInfo
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolversearchAllTweetsByKeyword
CloudFormation - CREATE_COMPLETE - AWS::DynamoDB::Table - UserTable
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolvermeInfo
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolversearchAllTweetsByKeyword
CloudFormation - CREATE_COMPLETE - AWS::AppSync::Resolver - GraphQlResolversearchAllTweetsByKeyword
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::DataSource - GraphQlDsUsers
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolvermeInfo
CloudFormation - CREATE_COMPLETE - AWS::AppSync::Resolver - GraphQlResolvermeInfo
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::DataSource - GraphQlDsUsers
CloudFormation - CREATE_COMPLETE - AWS::AppSync::DataSource - GraphQlDsUsers
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolvergetUserInfo
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolverupdateUserInfo
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolvergetUserInfo
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolverupdateUserInfo
CloudFormation - CREATE_COMPLETE - AWS::AppSync::Resolver - GraphQlResolvergetUserInfo
CloudFormation - CREATE_COMPLETE - AWS::AppSync::Resolver - GraphQlResolverupdateUserInfo
CloudFormation - CREATE_COMPLETE - AWS::DynamoDB::Table - TweetsTable
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::DataSource - GraphQlDsTweets
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::DataSource - GraphQlDsTweets
CloudFormation - CREATE_COMPLETE - AWS::AppSync::DataSource - GraphQlDsTweets
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolvertopTweet
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolverreTweet
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolverupdateTweet
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolverdeleteTweet
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolvertopTweet
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolvertweets
CloudFormation - CREATE_COMPLETE - AWS::AppSync::Resolver - GraphQlResolvertopTweet
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolverreTweet
CloudFormation - CREATE_COMPLETE - AWS::AppSync::Resolver - GraphQlResolverreTweet
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolverdeleteTweet
CloudFormation - CREATE_COMPLETE - AWS::AppSync::Resolver - GraphQlResolverdeleteTweet
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolvertweets
CloudFormation - CREATE_COMPLETE - AWS::AppSync::Resolver - GraphQlResolvertweets
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolvercreateTweet
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolvercreateTweet
CloudFormation - CREATE_COMPLETE - AWS::AppSync::Resolver - GraphQlResolvercreateTweet
CloudFormation - CREATE_IN_PROGRESS - AWS::AppSync::Resolver - GraphQlResolverupdateTweet
CloudFormation - CREATE_COMPLETE - AWS::AppSync::Resolver - GraphQlResolverupdateTweet
CloudFormation - UPDATE_COMPLETE_CLEANUP_IN_PROGRESS - AWS::CloudFormation::Stack - appsync-example-dev
CloudFormation - UPDATE_COMPLETE - AWS::CloudFormation::Stack - appsync-example-dev
Serverless: Stack update finished...
Service Information
service: appsync-example
stage: dev
region: us-west-2
stack: appsync-example-dev
api keys:
  None
endpoints:
  None
functions:
  graphql: appsync-example-dev-graphql

Stack Outputs
GraphQlApiUrl: https://XXX.appsync-api.us-west-2.amazonaws.com/graphql
GraphqlLambdaFunctionQualifiedArn: arn:aws:lambda:us-west-2:12345:function:appsync-example-dev-graphql:3
ServerlessDeploymentBucketName: appsync-example-dev-serverlessdeploymentbucket-XXX
```
And screenshots from the AWS console: https://imgur.com/a/N0HsvJg

### CODE REVIEW

**Comment by pvalentim** [comment]
Can we get this merged please :)

File: index.js [hunk 1]
```diff
@@ -78,420 +65,156 @@ class ServerlessAppsyncPlugin {
...
-      authenticationType: resolvedConfig.authenticationType,
-      name: resolvedConfig.name,
+    const resources = this.serverless.service.provider.compiledCloudFormationTemplate.Resources;
+
+    Object.assign(resources, this.getGraphQlApiEndpointResource(config));
+    Object.assign(resources, this.getApiKeyResources(config));
+    Object.assign(resources, this.getGraphQLSchemaResource(config));
+    Object.assign(resources, this.getDataSourceResources(config));
+    Object.assign(resources, this.getResolverResources(config));
+    Object.assign(resources, this.getResolverResources(config));
```
Comment (jdeanwaite) [inline index.js]: Is this line supposed to be duplicated?

File: index.js [hunk 2]
```diff
@@ -78,420 +65,156 @@ class ServerlessAppsyncPlugin {
...
+  getDataSourceResources(config) {
+    return config.dataSources.reduce((acc, ds) => {
+      const resource = {
+        Type: 'AWS::AppSync::DataSource',
+        Properties: {
+          ApiId: { 'Fn::GetAtt': ['GraphQlApi', 'ApiId'] },
+          Name: ds.name,
+          Description: ds.description,
+          Type: ds.type,
+          ServiceRoleArn: ds.config.serviceRoleArn,
```
Comment (jdeanwaite) [inline index.js]: ServiceRoleArn is not required for Local resolvers (those of type "NONE"). This causes the following error:
`TypeError: Cannot read property 'serviceRoleArn' of undefined`. 

```
      - type: NONE
        name: Local
        description: "Local data source for mapping fields."
```

**Comment by jdeanwaite** [comment]
@ajmath Thanks for the work on this plugin! I started using it and ran into a couple of problems.
1. Proper support should be implemented for data sources of type `NONE`. See my comment.
2. When installing this fresh I had this error:
`ENOENT: no such file or directory, open '/Users/justin/WebstormProjects/my-project/services/appsync/node_modules/serverless-appsync-plugin/example/.serverless_plugins/serverless-appsync-plugin'`

I went in and deleted the `.serverless_plugins` folder from the example directory and that resolved the issue.

I hope this can get merged soon! This offers a much more robust and simple way to develop an AppSync API.

**Comment by sid88in** [comment]
thanks @ajmath. I plan to review/merge this PR sometime next week. I'll test it once E2E to make sure everything is good. Overall looks great + good support :)

File: index.js (hunk 1 above)
Comment (ajmath) [reply index.js]: Nope.  Good catch.

**Comment by jdeanwaite** [comment]
Awesome, thanks for the updates! I started using this last night for a project I’m working on and it definitely fits the bill for what I was needing. Trying to use AppSync with straight cloud formation is quite the development nightmare. Server less makes this a lot easier. Thanks again for your work on this!

**Comment by guilhermedecampo** [comment]
Top notch PR! Please merge!

**Comment by austin43** [comment]
Great job with this PR! It seems to work with my current stack well.

I have a couple items of feedback:
* Is it possible to somehow retain update-appsync? The old method of deploy-appsync, update-appsync carried some significant deploy performance benefits over a whole stack deployment.  Now, when deploying it forces an upload of all of my lambda functions and other resources.  Until serverless-offline supports appsync, it may hinder development speed if we have to deploy everything.
* Support for aliases would be ideal for teams.  In the old method, one could deploy as many new schemas as they wanted.  With the new method, if it is limited to one appsync schema per stage, it may become difficult to develop with larger teams doing schema work in parallel.  https://github.com/hyperbrain/serverless-aws-alias

Overall I am on board with the idea of managing appsync via cfn, as it seems much cleaner.  Let me know if the above items make any sense and/or warrant separate issues to add support for what I mentioned.

**Comment by ajmath** [comment]
@austin43 Thanks for the feedback.
>Is it possible to somehow retain update-appsync?

I understand your frustration with increased deployment times but I don't think that updating the resources outside of CFN is a good idea.  It would certainly be possible but I have concerns about having multiple modifiers of the state of the app sync resources and this leading to unexpected inconsistencies.

> Support for aliases would be ideal for teams.

I don't think that multiple actors independently deploying to the same stage is a supported workflow.  To build in support for app sync aliases, I would suggest that those features go into a new plugin since it would be tightly coupling this plugin to [hyperbrain/serverless-aws-alias](https://github.com/hyperbrain/serverless-aws-alias).  As an alternative, maybe you could have each developer deploy to their own stage?

**Comment by guilhermedecampo** [comment]
yeah @ajmath for the second point we are using some naming convention of the appsync api like `nameproject-stage-developername` to make environments for developers. 
Seems to be ok for now.

**Comment by jdeanwaite** [comment]
@austin43 

I definitely agree with your points, however I am afraid most of those are likely limitations of using CloudFormation and are a little bit out of control of the author of this PR. However, I did want to point out that when you redeploy a CF stack, only the items that changed actually change in the stack, so it's not like you have to wait for every resource to delete and recreate. It still isn't the fastest, but in my experience (using heavily today) it only takes about 20-30 seconds (still is a lot, I know).

**Comment by jdeanwaite** [comment]
if you use yarn, 
`yarn add https://github.com/sid88in/serverless-appsync-plugin#91/head`

**Comment by austin43** [comment]
@jdeanwaite Ah I figured as much.  My point about CF resources "redeploying" was more along the lines of it has to waste time checking everything.  We may have to examine how to speed up our deploys in other ways then.

**Comment by austin43** [comment]
@ajmath I agree wholly with your first point, and with @jdeanwaite's comments it seems like we're limited by cloudformation and how it maintains stack consistency.

On your second point I think it's possible to use stages per dev, but this increases development costs as we have to deploy provisioned resources such as DynamoDB tables or Kinesis Streams.  We may just have to look forward to offline support!  Thank you both for your responses.

**Comment by ajmath** [comment]
@austin43 For offline support, my team has been maintaining two different entry points into the app, one using graphql-yoga to standup a local server and the other for exporting the lambda handler.  It will depend on how your project is setup to know if this will work for you.  We're using https://github.com/19majkel94/type-graphql to define the schema using decorators which makes the local setup a breeze.

**Comment by jdeanwaite** [comment]
@austin43 One other thing that has sped up deployment for me is if you are using Lambda functions are your data sources you can deploy individual functions (wayy faster). Also as my stack grew it definitely started taking a lot longer to deploy changes.

`sls deploy -f myResolverFn`

**Comment by guilhermedecampo** [comment]
Interesting idea @ajmath for offline support. Gonna dig into this graphql-yoga.

**Comment by sid88in** [comment]
@ajmath merging this! Great work! Thanks :) Looking forward to more. Also, I am going to make you a contributor of this repository. Cheers.

**Comment by sid88in** [comment]
@ajmath published 1.0.0 https://www.npmjs.com/package/serverless-appsync-plugin and added you as collaborator ;) cheers!

**Pull request merged**